### ISSUE TITLE
Custom volumes for each configuration of a matrix job

### ISSUE THREAD

**Issue by tashoyan** [jenkinsci/docker-plugin#734]
It would be good to configure volumes per job or even per matrix configuration

## Problem:
Several jobs use the same Docker cloud and the same Docker agent template. These jobs need Groovy grapes cache. To save time on downloading cached libs, this cache is mounted as a volume. This volume is configured in Configure System / Docker Cloud / Docker agent template configuration.
The physical location of this volume is the same for all jobs and all matrix configuration. This is not good, because we need separate Groovy Grapes caches for separate jobs and even matrix configurations.

Could it be possible to configure volumes using environment variables like JOB_NAME and/or custom variables provided by Matrix Project plugin? For example:
```
/data/ci/docker_volumes/cache-$JOB_NAME-$test_set:/build/.groovy/grapes:rw
```

**Comment by pjdarton** [comment]
No, this doesn't fit with how this plugin works; Jenkins asks for slaves matching labels and the plugin looks for pre-defined templates that match those labels and starts containers based on those templates.
When starting a template, this plugin has no knowledge of what build(s) that container will be used for.

You _could_ define a load of templates, all _very_ similar, differing only in the volumes which are labelled with job names and test sets, and then specify (in your matrix job) axes that cover all combinations.
Or, you could put a build step in your build (at the beginning) that symlinked /build/.groovy/grapes to somewhere within a single (large, common) volume (that held all the data for all your tests).

Or (and this is what I'd recommend) you could re-code your matrix builds as pipeline builds and use the docker-pipeline-plugin which will give you the flexibility you're looking for.
Matrix builds are ok for simple stuff (for _really_ simple stuff) but pipelines allow you to do everything that matrix builds do and way more.

TL;DR: Pipelines are cool; matrix builds are unfashionable ;-)

**Comment by tashoyan** [comment]
One more workaround: configure Groovy step in our matrix job to use custom locations for Grapes cache:
```bash
groovy -Dgrape.root=<location on the volume>/$JOB_NAME-$test_set
```
Link to the Groovy Grapes doc: [Repository Directory](http://docs.groovy-lang.org/latest/html/documentation/grape.html#Grape-RepositoryDirectory)

But your idea of migrating to pipelines seems even more attractive.
